src/plugins/presentation_util/public/components/floating_actions/floating_actions.tsx

```diff
@@ -41,59 +46,88 @@ export const FloatingActions: FC<FloatingActionsProps> = ({
+  const [floatingActions, setFloatingActions] = useState<FloatingActionItem[]>([]);
 
   useEffect(() => {
     if (!api) return;
 
-    const getActions = async () => {
-      let mounted = true;
-      const context = {
-        embeddable: api,
-        trigger: panelHoverTrigger,
-      };
+    let mounted = true;
+    const context = {
+      embeddable: api,
+      trigger: panelHoverTrigger,
+    };
+
+    const sortByOrder = (a: Action | FloatingActionItem, b: Action | FloatingActionItem) => {
+      return (a.order || 0) - (b.order || 0);
+    };
+
+    const getActions: () => Promise<FloatingActionItem[]> = async () => {
       const actions = (
         await uiActionsService.getTriggerCompatibleActions(PANEL_HOVER_TRIGGER, context)
       )
-        .filter((action): action is Action & { MenuItem: React.FC<{ context: unknown }> } => {
+        .filter((action) => {
           return action.MenuItem !== undefined && (disabledActions ?? []).indexOf(action.id) === -1;
         })
-        .sort((a, b) => (a.order || 0) - (b.order || 0));
+        .sort(sortByOrder);
+      return actions as FloatingActionItem[];
+    };
+
+    const subscriptions = new Subscription();
 
+    const handleActionCompatibilityChange = (isCompatible: boolean, action: Action) => {
       if (!mounted) return;
-      if (actions.length > 0) {
-        setFloatingActions(
-          <>
-            {actions.map((action) =>
-              React.createElement(action.MenuItem, {
-                key: action.id,
-                context,
-              })
-            )}
-          </>
+      setFloatingActions((currentActions) => {
+        const newActions: FloatingActionItem[] = currentActions
+          ?.filter((current) => current.id !== action.id)
+          .sort(sortByOrder) as FloatingActionItem[];
+        if (isCompatible) {
+          return [action as FloatingActionItem, ...newActions];
+        }
+        return newActions;
+      });
+    };
+
+    (async () => {
+      const actions = await getActions();
+      if (!mounted) return;
+      setFloatingActions(actions);
+
+      const frequentlyChangingActions = uiActionsService.getFrequentlyChangingActionsForTrigger(
+        PANEL_HOVER_TRIGGER,
+        context
+      );
+
+      for (const action of frequentlyChangingActions) {
+        subscriptions.add(
+          action.subscribeToCompatibilityChanges(context, handleActionCompatibilityChange)
         );
-      } else {
-        setFloatingActions(undefined);
       }
-      return () => {
-        mounted = false;
-      };
-    };
+    })();
 
-    getActions();
+    return () => {
+      mounted = false;
+      subscriptions.unsubscribe();
+    };
   }, [api, viewMode, disabledActions]);
 
   return (
     <div className="presentationUtil__floatingActionsWrapper">
       {children}
-      {isEnabled && floatingActions && (
+      {isEnabled && Boolean(floatingActions.length) && (
```

Comment:
I would also like to ask about this - I tweaked this condition because when there were no available actions, in the ui you would see this almost invisible white blank circle (border from the element), so here I am trying to make sure that the component is not rendered at all if there are no actions.
(When I initially added `isEnabled && floatingActions.length && (`
it would show "0" on the screen...
So I thought of 2 options:
1. `isEnabled && floatingActions.length > 0 && (`
2. `isEnabled && Boolean(floatingActions.length) && (`
Any preferences? 😊

![Screenshot 2024-11-20 at 18 53 00](https://github.com/user-attachments/assets/93d02413-49df-45cd-8202-58301daae7cb)

Reply:
I like `isEnabled && floatingActions.length > 0`, personally - it's just a little bit easier to read IMO :) But this is very much a style choice, so I am good with either option